server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -1702,6 +1709,11 @@ public static IndexMetadata readFrom(StreamInput in, @Nullable Function<String,
                 builder.putMapping(new MappingMetadata(in));
             }
         }
+        int inferenceFieldsSize = in.readVInt();
+        for (int i = 0; i < inferenceFieldsSize; i++) {
```

Comment:
Why not use `in.readCollection()` or `in.readMap()` here? We're assuming how `out.writeCollection` works instead, which seems a bit weird to me 🤔